EIPS/eip-4844.md

```diff
@@ -0,0 +1,558 @@
+    max_basefee_per_gas: uint256
+    gas: uint64
+    to: Union[None, Address] # Address = Bytes20
+    value: uint256
+    data: ByteList[MAX_CALLDATA_SIZE]
+    access_list: List[AccessTuple, MAX_ACCESS_LIST_SIZE]
+    blob_versioned_hashes: List[VersionedHash, MAX_VERSIONED_HASHES_LIST_SIZE]
+
+class AccessTuple(Container):
+    address: Address # Bytes20
+    storage_keys: List[Hash, MAX_ACCESS_LIST_STORAGE_KEYS]
+
+class ECDSASignature(Container):
+    v: uint8
+    r: uint256
+    s: uint256
+```
+
+The `priority_fee_per_gas` and `max_basefee_per_gas` fields follow [EIP-1559](./eip-1559.md) semantics,
+and `access_list` as in [`EIP-2930`](./eip-2930.md).
+
+[`EIP-2718`](./eip-2718.md) is extended with a "wrapper data", the typed transaction can be encoded in two forms, dependent on the context:
+- Network (default): `TransactionType || TransactionNetworkPayload`, or `LegacyTransaction`
+- Minimal (as in execution payload): `TransactionType || TransactionPayload`, or `LegacyTransaction`
+
+Execution-payloads / blocks use the minimal encoding of transactions.
+In the transaction-pool and local transaction-journal the network encoding is used.
+
+For previous types of transactions the network encoding is no different, i.e. `TransactionNetworkPayload == TransactionPayload`.
+
+The `TransactionNetworkPayload` wraps a `TransactionPayload` with additional data:
+this wrapping data SHOULD be verified directly before or after signature verification.
+
+When a blob transaction is passed through the network (see the [Networking](#Networking) section below),
+the `TransactionNetworkPayload` version of the transaction also includes `blobs` and `kzgs` (commitments list).
+The execution layer verifies the wrapper validity against the inner `TransactionPayload` after signature verification as:
+
+* All hashes in `blob_versioned_hashes` must start with the byte `BLOB_COMMITMENT_VERSION_KZG`
+* There may be at most `MAX_BLOBS_PER_TX` blob commitments in any single transaction.
+* There may be at most `MAX_BLOBS_PER_BLOCK` total blob commitments in a valid block.
+* There is an equal amount of versioned hashes, kzg commitments and blobs.
+* The KZG commitments hash to the versioned hashes, i.e. `kzg_to_versioned_hash(kzg[i]) == versioned_hash[i]`
+* The KZG commitments match the blob contents. (Note: this can be optimized with additional data, using a proof for a 
+  random evaluation at two points derived from the commitment and blob data)
+
+
+The signature is verified and `tx.origin` is calculated as follows:
+
+```python
+def tx_hash(tx: SignedBlobTransaction) -> Bytes32:
+    # The pre-image is prefixed with the transaction-type to avoid hash collisions with other tx hashers and types
+    return keccak256(BLOB_TX_TYPE + ssz.hash_tree_root(tx.message))
+
+def get_origin(tx: SignedBlobTransaction) -> Address:
+    sig = tx.signature
+    return ecrecover(tx_hash(tx), sig.v, sig.r, sig.s)
+```
+
+### Beacon chain validation
```

Comment:
Suggest moving CL code changes out into `consensus-specs` repo and use "events" and other types of mechanisms to just reference the outer CL functionality

Reply:
Will work on a draft for the CL side this week. More clarification about the consensus specs and EIP relation would be appreciated.